### PR TITLE
Various ports from Polaris and Bay, fluid updates.

### PR DESCRIPTION
**Pull request by Zuhayr**
- Fixes multiz pipes not working.
- Fixes multiz cabling not working.
- Reduces blocking lag from asteroid gen by something like three quarters, @Lohikar is a v. clever boy.
- Improves fluid controller performance via macros.
- Makes fluids Z-aware (fluids will now flow through open space beneath them, by preference, over their neighbors).
- Re-adds a deleted `water_act` block so lights and APCs and such will now be damaged by flooding.

### CODE REVIEW

File: code/modules/europa/fluids/fluid_turf.dm [hunk 1]
```diff
@@ -22,29 +22,25 @@
...
+	if(F) LOSE_FLUID(F, amount)
 
 /turf/return_fluid()
 	return (locate(/obj/effect/fluid) in contents)
 
 /turf/Destroy()
 	fluid_update()
-	if(fluid_master)
-		fluid_master.remove_active_source(src)
+	if(SSfluids)
```
Comment (Lohikar) [inline code/modules/europa/fluids/fluid_turf.dm]: No point checking if SSfluids exists, it should *always* exist (initialized with globals, before world/New).

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 2]
```diff
@@ -170,7 +170,7 @@ meteor_act
 	if(!affecting)
 		return //should be prevented by attacked_with_item() but for sanity.
 
-	visible_message("<span class='danger'>[src] has been [I.attack_verb.len? pick(I.attack_verb) : "attacked"] in the [affecting.name] with [I.name] by [user]!</span>")
+	visible_message("<span class='danger'>[src] has been [(I.attack_verb && I.attack_verb.len) ? pick(I.attack_verb) : "attacked"] in the [affecting.name] with [I.name] by [user]!</span>")
```
Comment (Lohikar) [inline code/modules/mob/living/carbon/human/human_defense.dm]: alternately: `[LAZYLEN(I.attack_verb) ? pick(I.attack_verb) : "attacked"]`
There might be a macro for this (LAZYPICK), not sure if it was brought over.

File: code/modules/random_map/automata/caves.dm [hunk 3]
```diff
@@ -2,8 +2,95 @@
...
+
+/datum/random_map/automata/asteroids/cave_system/apply_to_map()
+	if(!origin_x) origin_x = 1
+	if(!origin_y) origin_y = 1
+	if(!origin_z) origin_z = 1
+
+	var/tmp_cell
+	var/new_path
+	var/num_applied = 0
+	for (var/thing in block(locate(origin_x, origin_y, origin_z), locate(limit_x, limit_y, origin_z)))
```
Comment (Lohikar) [inline code/modules/random_map/automata/caves.dm]: Shouldn't this be `origin_x + limit_x`?

File: code/__defines/fluids.dm [hunk 4]
```diff
@@ -5,12 +5,12 @@
 #define FLUID_MAX_DEPTH FLUID_DEEP*4       // Arbitrary max value for flooding.
 
 // Expects /turf for T.
-#define ADD_ACTIVE_FLUID_SOURCE(T)    if(SSfluids) SSfluids.water_sources |= T
-#define REMOVE_ACTIVE_FLUID_SOURCE(T) if(SSfluids) SSfluids.water_sources -= T
+#define ADD_ACTIVE_FLUID_SOURCE(T)    SSfluids.water_sources |= T
```
Comment (Lohikar) [inline code/__defines/fluids.dm]: Might be worth making this assoc instead of `|=`, though it's prob. fine.

File: code/__defines/fluids.dm [hunk 5]
```diff
@@ -5,12 +5,12 @@
...
+#define REMOVE_ACTIVE_FLUID_SOURCE(T) SSfluids.water_sources -= T
 
 // Expects /obj/effect/fluid for F.
-#define ADD_ACTIVE_FLUID(F)           if(SSfluids) SSfluids.active_fluids |= F
-#define REMOVE_ACTIVE_FLUID(F)        if(SSfluids) SSfluids.active_fluids -= F
+#define ADD_ACTIVE_FLUID(F)           SSfluids.active_fluids |= F
```
Comment (Lohikar) [inline code/__defines/fluids.dm]: Same here.

File: code/__defines/fluids.dm [hunk 6]
```diff
@@ -54,3 +54,9 @@
 		} \
 	} \
 	if(!flooded_a_neighbor) REMOVE_ACTIVE_FLUID_SOURCE(T);
+
+// We share overlays for all fluid turfs to sync icon animation.
+var/list/fluid_images = list()
+#define APPLY_FLUID_OVERLAY(img_state) \
+	if(!fluid_images[img_state]) fluid_images[img_state] = image('icons/effects/liquids.dmi',img_state); \
```
Comment (Lohikar) [inline code/__defines/fluids.dm]: Not really any benefit to using images if you're just leaving them as default anyways.

Comment (Zuhayr) [reply code/__defines/fluids.dm]: Going to leave it explicit so that the base object can use a different icon without messing it up.

**Pull request merged**